hbase-server/src/main/java/org/apache/hadoop/hbase/client/ClientSideRegionScanner.java

```diff
@@ -60,6 +62,14 @@ public ClientSideRegionScanner(Configuration conf, FileSystem fs,
     region = HRegion.newHRegion(CommonFSUtils.getTableDir(rootDir, htd.getTableName()), null, fs,
       conf, hri, htd, null);
     region.setRestoredRegion(true);
+    // non RS process does not have a block cache, and this a client side scanner,
+    // create one for MapReduce jobs to cache the INDEX block
+    conf.setFloat(HConstants.HFILE_BLOCK_CACHE_SIZE_KEY,
+      conf.getFloat(HConstants.HBASE_CLIENT_SCANNER_BLOCK_CACHE_SIZE_KEY,
+        HConstants.HBASE_CLIENT_SCANNER_BLOCK_CACHE_SIZE_DEFAULT));
+    // don't allow L2 bucket cache for non RS process to avoid unexpected disk usage.
+    conf.unset(HConstants.BUCKET_CACHE_IOENGINE_KEY);
```

Comment:
[DISCUSS] this `ClientSideRegionScanner` should be only used by downstream client, e.g. MR input format like `TableSnapshotInputFormatImpl` and `TableSnapshotScanner`  , such that we should not allow it to create additional disk cache. IMO we should disable it at least before any one has a use case for it, and avoid unexpected resource usage.

Reply:
Oh ya.. I came to this PR to check this aspect.  I dont think it will be good if we allow to create offheap or file based BucketCache as part of MR scans.

Reply:
this line should block the use of file based BucketCache, but I didn't unset the key of `EXTERNAL_BLOCKCACHE_KEY`/`hbase.blockcache.external.class` which should be default to off.

Reply:
agree with Anoop -- no need for anything but an L1 on-heap blockcache (and specifically L1 only for index blocks, not data blocks: how we normally do it when we use bucketcache).

However, because we unset the bucketcache, that means that we don't use CombinedBlockCache, but only LruBlockCache directly in the ClientSideRegionScanner. This means that the block cache will also be caching data blocks which I don't think was the spirit of this change. The problem is that I don't see a way for LruBlockCache to only cache index blocks.

I feel like we'd have to update `CombinedBlockCache` to allow for a null `L2` and make a special BlockCacheFactory call for this special, client-side case.

Reply:
> have to update CombinedBlockCache to allow for a null L2 and make a special BlockCacheFactory call for this special, client-side case.

good point, let me rework on it.

Reply:
@joshelser  @anoopsjohn   at the end I decided to add a new L1 cache `IndexOnlyLruBlockCache` that only need `cacheBlock` verify if the block is not a data block.

Reply:
It looks great to me